### PR TITLE
Adds JSON extension data to labeled utterance model

### PR DESCRIPTION
**Pull request by rozele**
In order to support a number of additional properties, including timestamps, intent confidence scores, transcription confidence scores, and utterance IDs, this change removes all specializations of NLU.DevOps.Models.LabeledUtterance in favor of NLU.DevOps.Core.JsonLabeledUtterance, which simply stores any additional properties in a dictionary labeled with JsonExtensionData from Json.NET.

A side-effect of this PR is that each NLU provider now labels utterances with a timestamp.

Towards #204

### CODE REVIEW

File: src/NLU.DevOps.Dialogflow/DialogflowNLUTestClient.cs [hunk 1]
```diff
@@ -73,9 +73,11 @@ protected override async Task<LabeledUtterance> TestAsync(string utterance, Canc
                     {
                         var client = await this.GetSessionClientAsync(cancellationToken).ConfigureAwait(false);
                         var result = await client.DetectIntentAsync(sessionName, queryInput, cancellationToken).ConfigureAwait(false);
-                        return new LabeledUtterance(
+                        return new PredictedLabeledUtterance(
                             result.QueryResult.QueryText,
                             result.QueryResult.Intent.DisplayName,
+                            result.QueryResult.IntentDetectionConfidence,
```
Comment (rozele) [inline src/NLU.DevOps.Dialogflow/DialogflowNLUTestClient.cs]: Update tests

File: src/NLU.DevOps.Lex/LexNLUTestClient.cs [hunk 2]
```diff
@@ -94,7 +94,7 @@ protected override async Task<LabeledUtterance> TestAsync(string utterance, Canc
                 .Select(slot => new Entity(slot.Key, slot.Value, null, 0))
                 .ToArray();
 
-            return new LabeledUtterance(
+            return new PredictedLabeledUtterance(
```
Comment (rozele) [inline src/NLU.DevOps.Lex/LexNLUTestClient.cs]: Update tests

File: src/NLU.DevOps.CommandLine/NLU.DevOps.CommandLine.csproj [hunk 3]
```diff
@@ -16,8 +16,7 @@
     <TargetsForTfmSpecificContentInPackage>PublishProviders;$(TargetsForTfmSpecificContentInPackage)</TargetsForTfmSpecificContentInPackage>
   </PropertyGroup>
 
-  <Target Name="PublishProviders"
-          BeforeTargets="CoreBuild">
+  <Target Name="PublishProviders" BeforeTargets="CoreBuild">
```
Comment (rozele) [inline src/NLU.DevOps.CommandLine/NLU.DevOps.CommandLine.csproj]: Revert.

File: src/NLU.DevOps.Core/LabeledUtterancePropertyExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        }
+
+        /// <summary>
+        /// Gets the timestamp for the labeled utterance.
+        /// </summary>
+        /// <param name="utterance">Labeled utterance.</param>
+        /// <returns>
+        /// Timestamp, or <code>null</code> if property is not set.
+        /// </returns>
+        public static DateTimeOffset? GetTimestamp(this LabeledUtterance utterance)
```
Comment (rozele) [inline src/NLU.DevOps.Core/LabeledUtterancePropertyExtensions.cs]: Move the getters to test code / ModelPerformance

File: src/NLU.DevOps.Dialogflow/DialogflowNLUTestClient.cs [hunk 5]
```diff
@@ -74,9 +74,12 @@ protected override async Task<LabeledUtterance> TestAsync(string utterance, Canc
...
                         return new LabeledUtterance(
-                            result.QueryResult.QueryText,
-                            result.QueryResult.Intent.DisplayName,
-                            result.QueryResult.Parameters?.Fields.SelectMany(GetEntities).ToList());
+                                result.QueryResult.QueryText,
+                                result.QueryResult.Intent.DisplayName,
+                                result.QueryResult.Parameters?.Fields.SelectMany(GetEntities).ToList())
+                            .WithScore(result.QueryResult.IntentDetectionConfidence)
+                            .WithTextScore(result.QueryResult.SpeechRecognitionConfidence)
+                            .WithTimestamp(DateTimeOffset.Now);
```
Comment (rozele) [inline src/NLU.DevOps.Dialogflow/DialogflowNLUTestClient.cs]: Should we bubble the `WithTimestamp` up to a different abstraction level (i.e., so each service doesn't have to do this individually)?

Comment (rozele) [reply src/NLU.DevOps.Dialogflow/DialogflowNLUTestClient.cs]: Maybe in a future PR.

File: src/NLU.DevOps.Core/LabeledUtterancePropertyExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+namespace NLU.DevOps.Core
+{
+    using System;
+    using Models;
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// Enable standard data extensions for <see cref="LabeledUtterance"/>.
+    /// </summary>
+    public static class LabeledUtterancePropertyExtensions
```
Comment (rozele) [inline src/NLU.DevOps.Core/LabeledUtterancePropertyExtensions.cs]: Add unit tests

**Comment by saroup** [comment]
Can you also update documentation to account for these changes?

File: src/NLU.DevOps.Luis/LuisNLUTestClient.cs [hunk 7]
```diff
@@ -140,17 +138,17 @@ Entity getEntity(EntityModel entity)
...
-            return !score.HasValue && Math.Abs(speechLuisResult.TextScore) < Epsilon
-                ? new LabeledUtterance(speechLuisResult.LuisResult.Query, intent, entities)
-                : new ScoredLabeledUtterance(speechLuisResult.LuisResult.Query, intent, score ?? 0, speechLuisResult.TextScore, entities);
+            return new LabeledUtterance(
+                    speechLuisResult.LuisResult.Query,
+                    speechLuisResult.LuisResult.TopScoringIntent?.Intent,
+                    speechLuisResult.LuisResult.Entities?.Select(getEntity).ToList())
+                .WithScore(speechLuisResult.LuisResult.TopScoringIntent?.Score)
+                .WithTextScore(speechLuisResult.TextScore)
+                .WithTimestamp(DateTimeOffset.Now);
```
Comment (saroup) [inline src/NLU.DevOps.Luis/LuisNLUTestClient.cs]: >.WithTimestamp(DateTimeOffset.Now); [](start = 16, length = 35)

does it have to be attached to each utterance?  We could have a timestamp at the beginning of the json file for all utterances

Comment (rozele) [reply src/NLU.DevOps.Luis/LuisNLUTestClient.cs]: Worried about long running tests that take O(1hr) where you may have trouble with hours / minutes in the future (I.e., book me a car in 15 minutes)

Comment (rozele) [reply src/NLU.DevOps.Luis/LuisNLUTestClient.cs]: Won't fix for now, we can optimize later.

Comment (rozele) [reply src/NLU.DevOps.Luis/LuisNLUTestClient.cs]: We may want to consider making the timestamp opt-in though...

**Comment by rozele** [comment]
> Can you also update documentation to account for these changes?

Done.

File: src/NLU.DevOps.Luis/LuisNLUTestClient.cs (hunk 7 above)
Comment (rozele) [reply src/NLU.DevOps.Luis/LuisNLUTestClient.cs]: (will send a follow up PR).

**Pull request merged**